sqlmesh/core/plan/definition.py

```diff
@@ -451,30 +452,44 @@ def has_unmodified_unpromoted(self) -> bool:
         )
 
     def _add_restatements(self, restate_models: t.Iterable[str]) -> None:
-        for table in restate_models:
-            downstream = self._dag.downstream(table)
-            if table in self.context_diff.snapshots:
-                downstream.append(table)
-
-            snapshots = self.context_diff.snapshots
-            downstream = [
-                d for d in downstream if not snapshots[d].is_symbolic and not snapshots[d].is_seed
-            ]
-
-            if not self.is_dev:
-                models_with_disabled_restatement = [
-                    f"'{d}'" for d in downstream if snapshots[d].model.disable_restatement
-                ]
-                if models_with_disabled_restatement:
-                    raise PlanError(
-                        f"Restatement is disabled for models: {', '.join(models_with_disabled_restatement)}."
-                    )
-
-            if not downstream:
+        def is_restateable_snapshot(snapshot: Snapshot) -> bool:
+            if not self.is_dev and snapshot.model.disable_restatement:
+                logger.debug("Restatement is disabled for model '%s'.", snapshot.name)
+                return False
+            return not snapshot.is_symbolic and not snapshot.is_seed
+
+        dummy_interval = (sys.maxsize, -sys.maxsize)
+        snapshot_mapping = self.context_diff.snapshots
+        restatements = {}
+        # Add restate snapshots and their downstream snapshots
+        for snapshot_name in restate_models:
+            if snapshot_name not in snapshot_mapping or not is_restateable_snapshot(
+                snapshot_mapping[snapshot_name]
+            ):
                 raise PlanError(
-                    f"Cannot restate from '{table}'. Either such model doesn't exist, no other materialized model references it, or restatement was disabled for this model."
+                    f"Cannot restate from '{snapshot_name}'. Either such model doesn't exist, no other materialized model references it, or restatement was disabled for this model."
                 )
-            self._restatements.update(downstream)
+            restatements[snapshot_name] = dummy_interval
+            for downstream_snapshot_name in self._dag.downstream(snapshot_name):
+                if is_restateable_snapshot(snapshot_mapping[downstream_snapshot_name]):
+                    restatements[downstream_snapshot_name] = dummy_interval
+        # Get restatement intervals for all restated snapshots and make sure that if a snapshot expands it's
+        # restatement range that it's downstream dependencies all expand their restatement ranges as well.
+        for snapshot_name in self._dag:
```

Comment:
can this be just done in one for loop? why do you loop twice? because you already traverse the downstream

Reply:
I believe the challenge with doing that is that the first loop is on `restate_models` which can be arbitrarily ordered. The algorithm relies on having nodes push the largest interval range down to their children and visiting the graph in an arbitrary order makes that more confusing. I was looking at alternative approaches but it became harder to reason about the implementation.

I did realize though that I don't need to check the full upstream but rather just the immediate parents of a given node. Also I skip traversing graph at all if there are no restatements. https://github.com/TobikoData/sqlmesh/pull/1322/commits/8776e6427a44510594df54586ba0569f1a46f1ce